tests/unit/utils/test_utils.py

```diff
@@ -878,3 +878,36 @@ def test_parse_duration_hypothesis(duration):
 ])
 def test_mimetype_extension(mimetype, extension):
     assert utils.mimetype_extension(mimetype) == extension
+
+
+class TestCleanupFileContext:
+
+    def test_no_file(self, tmp_path):
+        tmpfile = tmp_path / 'tmp.txt'
+        with utils.cleanup_file(tmpfile):
+            pass
+        assert not tmpfile.exists()
+
+    def test_no_error(self, tmp_path):
+        tmpfile = tmp_path / 'tmp.txt'
+        with tmpfile.open('w'):
+            pass
+        with utils.cleanup_file(tmpfile):
+            pass
+        assert not tmpfile.exists()
+
+    def test_error(self, tmp_path):
+        tmpfile = tmp_path / 'tmp.txt'
+        with tmpfile.open('w'):
+            pass
+        with pytest.raises(RuntimeError):
+            with utils.cleanup_file(tmpfile):
+                raise RuntimeError
+        assert not tmpfile.exists()
+
+    def test_directory(self, tmp_path, caplog):
+        assert tmp_path.is_dir()
+        # removal of file fails since it's a directory
+        with caplog.at_level(logging.ERROR, 'misc'):
+            with utils.cleanup_file(tmp_path):
+                pass
```

Comment:
As for your question regarding checking the log message: See the [caplog documentation](https://docs.pytest.org/en/stable/logging.html#caplog-fixture) (and perhaps also [API reference](https://docs.pytest.org/en/stable/reference.html#pytest.LogCaptureFixture).

You'll need to be careful because the exact error message will be different on Linux/Windows. Perhaps something like this would suffice:

```python
assert len(caplog.messages) == 1
assert caplog.messages[0].startswith("Failed to delete tempfile")
```

Reply:
Added this and also for the test when there is no such file